ext/crates/fp/src/vector/mod.rs

```diff
@@ -97,92 +95,192 @@ mod test {
         }
     }
 
-    fn random_vector(p: impl Into<u32>, dimension: usize) -> Vec<u32> {
-        let p: u32 = p.into();
-        let mut result = Vec::with_capacity(dimension);
+    fn random_vector(p: u32, dimension: usize) -> Vec<u32> {
         let mut rng = rand::thread_rng();
-        for _ in 0..dimension {
-            result.push(rng.gen::<u32>() % p);
+        (0..dimension).map(|_| rng.gen_range(0..p)).collect()
+    }
+
+    /// An arbitrary `ValidPrime`
+    fn arb_prime() -> impl Strategy<Value = ValidPrime> {
+        prop_oneof![
+            Just(ValidPrime::new(2)),
+            Just(ValidPrime::new(3)),
+            Just(ValidPrime::new(5)),
+            Just(ValidPrime::new(7)),
+        ]
+    }
+
+    /// An arbitrary number of dimensions. This makes the functions that create vectors DRY.
+    fn arb_dim() -> impl Strategy<Value = usize> {
+        0usize..=10_000
+    }
+
+    /// An arbitrary pair of dimensions, where the second dimension is bigger than or equal to the
+    /// first. This happens to be the same as `arb_slice(10000)`, but we make it a separate function
+    /// because the semantics are different.
+    fn arb_dim_pair() -> impl Strategy<Value = (usize, usize)> {
+        arb_slice(10_000)
+    }
+
+    /// The start and end positions of an arbitrary slice of a vector of length `dimension`
+    fn arb_slice(dimension: usize) -> impl Strategy<Value = (usize, usize)> {
+        // We have every integer twice because `subsequence` returns distinct values, but we want to
+        // also test the case where the slices have the same start and end positions.
+        let all_indices: Vec<_> = (0..=dimension).flat_map(|i| [i, i]).collect();
+        proptest::sample::subsequence(all_indices, 2).prop_map(|v| (v[0], v[1]))
+    }
+
+    /// An arbitrary pair of slices of a vector of length `dimension` _that have the same length_
+    fn arb_slice_pair(dimension: usize) -> impl Strategy<Value = [(usize, usize); 2]> {
+        // Similarly to `arb_slice`, we triplicate the entries because we want to also test the case
+        // where two or all three values coincide.
+        let all_indices: Vec<_> = (0..=dimension).flat_map(|i| [i, i, i]).collect();
+        proptest::sample::subsequence(all_indices, 3)
+            .prop_map(|v| [(v[0], v[1]), (v[2] - (v[1] - v[0]), v[2])])
+            .prop_shuffle()
+    }
+
+    /// An arbitrary vector of length `dimension` containing values in the range `0..p`
+    fn arb_vec_u32(p: ValidPrime, dimension: usize) -> impl Strategy<Value = Vec<u32>> {
+        proptest::collection::vec(0..*p, dimension)
+    }
```

Comment:
I had the same idea, but we need the original `Vec<u32>` because many of the tests will apply some operation on both the `FpVector` and the original `Vec` and then compare the results.

Reply:
Makes sense, please add a comment.